Missing fmt::Debug impls for public structs and enums

lopopolo opened the pull request:
The Rust API guidelines advise that [all public types implement common `std` traits for interoperability](https://rust-lang.github.io/api-guidelines/interoperability.html), of which `fmt::Debug` is one.

This PR adds `fmt::Debug` implementations for all public types in `regex` and `regex-syntax` crates. This PR also adds `#![warn(missing_debug_implementations)]` to `lib.rs` in each crate.

This PR also opportunistically implements `Clone` on iterators that wrap `slice::Iter`.

Context: I'm building an implementation of Ruby `Regexp` on top of the `regex` crate, and having all public types implement `fmt::Debug` will allow me to `#[derive(Debug)]` on types that wrap e.g. iterators in this crate.

BurntSushi commented on src/dfa.rs:
```diff
@@ -279,6 +279,7 @@ fn push_inst_ptr(data: &mut Vec<u8>, prev: &mut InstPtr, ip: InstPtr) {
     *prev = ip;
 }
 
+#[derive(Debug)]
```
I don't think this is needed?

BurntSushi commented on src/re_trait.rs:
```diff
@@ -172,6 +174,21 @@ where
+            .field("last_match", &self.last_match)
+            .finish()
+    }
+}
```
I suspect you're writing out this impl because `Matches` is generic and doesn't require its type parameters to impl `Debug`? The `RegularExpression` trait is an internal construct, so adding a requirement to impl `Debug` seems fine to me. (For both `Self` and for its `Text` associated type.) Then you should be able to derive `Debug` impls here.

lopopolo replied:
I also made the same change to `CaptureMatches`.

BurntSushi commented on src/exec.rs:
```diff
@@ -106,6 +109,7 @@ pub struct ExecBuilder {
 
 /// Parsed represents a set of parsed regular expressions and their detected
 /// literals.
+#[derive(Debug)]
```
Is this `Debug` impl necessary?

BurntSushi commented on src/exec.rs:
```diff
@@ -97,6 +99,7 @@ struct ExecReadOnly {
 /// Facilitates the construction of an executor by exposing various knobs
 /// to control how a regex is executed and what kinds of resources it's
 /// permitted to use.
+#[derive(Debug)]
```
Is this `Debug` impl necessary?

lopopolo replied:
`ExecBuilder` is public via the `internal` module which "exists to support suspicious activity".

I'll remove the debug impl and allow the lint.

BurntSushi commented on src/compile.rs:
```diff
@@ -25,6 +26,7 @@ struct Patch {
 
 /// A compiler translates a regular expression AST to a sequence of
 /// instructions. The sequence of instructions represents an NFA.
+#[derive(Debug)]
```
Is this `Debug` impl necessary?

lopopolo replied:
`Compiler` is public via the internal module which "exists to support suspicious activity".

I'll remove the debug impl and allow the lint.

BurntSushi replied:
Haha right. Yeah, it's technically public, but is `doc(hidden)` and not a stable part of the API. I'm hoping to remove that quirk in the move to `regex-automata`.

Thanks.

BurntSushi reviewed (approved):
LGTM, but like the other PR, my sense is that some of these impls aren't needed to pass the lint? Or does the lint apply to internal types too?

The reason why I'm being a stickler about this is that I'd rather not add impls for things that we don't need because they likely increase compile times. And the compile times for this crate are already pretty poor.

BurntSushi reviewed (changes requested):
Whoops, I meant to "request changes."

lopopolo commented:
@BurntSushi I've made the requested changes. I squashed these into the commit for the `regex` crate.

For types that are public via the hidden `internal` module, I've added an `#[allow(missing_debug_implementations)]` and avoid implementing `Debug`.

lopopolo commented:
rebase incoming.

lopopolo commented:
> LGTM, but like the other PR

For reference, the other PR is #734.

BurntSushi reviewed (approved):
OK, LGTM, thanks!

lopopolo commented:
Whoops looks like I accidentally didn't split out the iterator traits bits from `regex-syntax` properly when splitting apart the two PRs.

Thanks for merging.

Would it be possible to get these two PRs packaged into a release soon?

BurntSushi commented:
Yeah I saw that, but that's okay. No biggie.

I'll try to do a release in the next couple days. Please ping me again if it doesn't happen.

lopopolo commented:
hi @BurntSushi following up with a friendly ping for a release.

BurntSushi commented:
@lopopolo OK, `regex 1.4.3` is on crates.io.

lopopolo commented:
Thank you!
